### PR TITLE
Add goja tools

### PR DESCRIPTION
**Pull request by rshopa**
Additional scripts for the transformation of the data added to a separate directory: goja-tools

### CODE REVIEW

**Comment by rshopa** [comment]
Hi guys,
Want to remind about this pull request - smearing hits and python scripts for remapping (3-layer into 1-layer)
Long time remained unresolved so far...

File: goja-tools/j-pet-layers-controller/src/LayersController.py [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+            'transformed': False
+        }
+
+    def add_zero_layer(self, total_strips_no=192):
+        """
+        Detects the radius which the strips are tightly composed in
+        :param total_strips_no: number of strips (default is 48+48+96)
+        :return: None
+        """
+        # (0.7/(2 sin(dQ/2))) * cos(dQ/2) + 0.95
```
Comment (wkrzemien) [inline goja-tools/j-pet-layers-controller/src/LayersController.py]: what is the relation between the comment and the code?

File: goja-tools/smear-events/src/SmearEvents.h [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+#include <math.h>       /* log, pow etc */
+#include <random>
+
+
+#ifndef SMEAREVENTS_H
+#define SMEAREVENTS_H
+// C_SCIN is the effective propagation velocity
+// of the signal along the strip [cm/ns]
+#define C_SCIN 12.6
```
Comment (wkrzemien) [inline goja-tools/smear-events/src/SmearEvents.h]: static const double C_SCIN=12.6

File: goja-tools/smear-events/src/SmearEvents.h [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    }
+    ~SmearEventsController(){}
+
+    // main function - imports data, smears it & exports
+    void blurData(const std::string &inputFileName,
+                  const std::string &outputFileName){
+        // stream
+        std::ifstream inputFile(inputFileName);
+        inputFile.clear();
+        inputFile.seekg(0, std::ios::beg); // move to top
```
Comment (wkrzemien) [inline goja-tools/smear-events/src/SmearEvents.h]: Why do you need those lines 69 and 70?

File: goja-tools/j-pet-layers-controller/src/LayersController.py (hunk 1 above)
Comment (rshopa) [reply goja-tools/j-pet-layers-controller/src/LayersController.py]: Removed

File: goja-tools/smear-events/src/SmearEvents.h (hunk 3 above)
Comment (rshopa) [reply goja-tools/smear-events/src/SmearEvents.h]: Removed but not tested directly in this branch. cksums of smeared files are equal on local node for old and new versions.

File: goja-tools/smear-events/src/SmearEvents.h (hunk 2 above)
Comment (rshopa) [reply goja-tools/smear-events/src/SmearEvents.h]: Changed but not tested directly in this branch. cksums of smeared files are equal on local node for old and new versions.

**Comment by flukson** [comment]
For me this PR is OK.
@graag @wkrzemien Do you have any other remarks? If not, could you please merge this PR?

File: goja-tools/smear-events/src/SmearEvents.h [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+        // --- The LOOP ---
+        // one row from data file:
+        std::string textRow;
+        // set random generator
+        RandomGenerator randGen(m_sigmas);
+        // iterator & event counter
+        int i(0);
+        std::stringstream s_EventNumber;
+
```
Comment (wkrzemien) [inline goja-tools/smear-events/src/SmearEvents.h]: I think you can define m_added_blur as local variable. It is used only inside this loop.
Vector added_blur{0.,0.,0.,0.}; probably would be enough

Comment (wkrzemien) [reply goja-tools/smear-events/src/SmearEvents.h]: The same with m_sigmas

Comment (rshopa) [reply goja-tools/smear-events/src/SmearEvents.h]: > The same with m_sigmas

Disagree. This parameter depends on scanner length (i.e. on the parameter ```scanner_length_ID``` set in the constructor). I may eliminate ```m_sigmas```
and replace it by some private ```m_scanner_length_ID```, but it makes no sense. Besides, possible future transformation of the script might need further operations with sigmas, e.g. whether to mix WLS with SiPM, PMT.

File: goja-tools/smear-events/src/SmearEvents.h [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    }
+};
+
+// A class for the manipulation of input data line by line
+class SmearEventsController{
+public:
+    // constructor and destructor
+    SmearEventsController(const std::string &readout,
+                          const Readout &readoutHash,
+                          const int &scanner_length_ID = 1)
```
Comment (wkrzemien) [inline goja-tools/smear-events/src/SmearEvents.h]: small thing - you dont need const and & with bultin types like int or double. The cost of copying by value is negligible

File: goja-tools/smear-events/src/SmearEvents.h (hunk 4 above)
Comment (wkrzemien) [reply goja-tools/smear-events/src/SmearEvents.h]: ok. Fair point. But m_added_blur can be.

Comment (wkrzemien) [inline goja-tools/smear-events/src/SmearEvents.h]:
```suggestion
Vector addedBlur{0.,0.,0.,0.};
```

File: goja-tools/smear-events/src/SmearEvents.h [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        // iterator & event counter
+        int i(0);
+        std::stringstream s_EventNumber;
+
+        while(std::getline(inputFile, textRow)){
+            i += 1;
+            // parse into values
+            Vector dataRow(m_import_text_event(textRow));
+            for(int i=0; i<2; i++){
+                m_addedBlur[i*2] = randGen.distributionZ(randGen.generator);
```
Comment (wkrzemien) [inline goja-tools/smear-events/src/SmearEvents.h]:
```suggestion
               addedBlur[i*2] = randGen.distributionZ(randGen.generator);
```

File: goja-tools/smear-events/src/SmearEvents.h [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+                m_addedBlur[i*2+1] = randGen.distributionTime(randGen.generator);
```
Comment (wkrzemien) [inline goja-tools/smear-events/src/SmearEvents.h]:
```suggestion
                addedBlur[i*2+1] = randGen.distributionTime(randGen.generator);
```

File: goja-tools/smear-events/src/SmearEvents.h [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+            }
+            // blur the data
+            Vector dataBlurred = m_blur_hits_and_times(dataRow,
+                                                       m_addedBlur);
```
Comment (wkrzemien) [inline goja-tools/smear-events/src/SmearEvents.h]:
```suggestion
                                 Vector dataBlurred = m_blur_hits_and_times(dataRow,
                                                       addedBlur);
```

Comment (wkrzemien) [inline goja-tools/smear-events/src/SmearEvents.h]:
```suggestion
                                                       addedBlur);
```

File: goja-tools/smear-events/src/SmearEvents.h [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+            std::cout << s_EventNumber.str() << std::flush << "\r";
+        }
+        outputFile.close();
+        std::cout << "Events exported: " << i << std::endl;
+    }
+
+private:
+    std::string m_readout;
+    Vector m_sigmas,
+           m_addedBlur; // 4-dimensional vector (for Z1,t1,Z2,t2)
```
Comment (wkrzemien) [inline goja-tools/smear-events/src/SmearEvents.h]:
```suggestion
```

File: goja-tools/smear-events/src/SmearEvents.h [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+            s_EventNumber << i;
+            std::cout << s_EventNumber.str() << std::flush << "\r";
+        }
+        outputFile.close();
+        std::cout << "Events exported: " << i << std::endl;
+    }
+
+private:
+    std::string m_readout;
+    Vector m_sigmas,
```
Comment (wkrzemien) [inline goja-tools/smear-events/src/SmearEvents.h]:
```suggestion
    Vector m_sigmas;
```

File: goja-tools/smear-events/src/SmearEvents.h (hunk 4 above)
Comment (rshopa) [reply goja-tools/smear-events/src/SmearEvents.h]: m_added_blur is now a local variable (see the latest commits).

File: goja-tools/smear-events/src/SmearEvents.h (hunk 5 above)
Comment (rshopa) [reply goja-tools/smear-events/src/SmearEvents.h]: Changed two lines for int and unsigned int without const and & (see the latest commit)

**Pull request merged**